mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/options/MongoDBIndexOptions.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.datastore.mongodb.options;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+/**
+ * MongoDB specific options for a given index.
+ *
+ * See <a href="https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/#options-for-all-index-types">https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/#options-for-all-index-types</a>
+ *
+ * @author Guillaume Smet
+ */
+@Target( {TYPE, METHOD, FIELD } )
+@Retention(RUNTIME)
+public @interface MongoDBIndexOptions {
+
+	/**
+	 * Name of the index to which these options apply.
+	 */
+	String forIndex();
+
+	/**
+	 * Builds the index in the background so that building an index does not block other database activities.
+	 */
+	boolean background() default false;
+
+	/**
+	 * Allows to define a partial index by providing a filter expression.
+	 *
+	 * Must be a valid MongoDB document.
+	 */
+	String partialFilterExpression() default "";
```

Comment:
Wow, this goes way beyond what we had in originally in mind. At least I just had thought about enabling usage of the existing JPA annotations for index creation. It ties into the discussion above really. I'm not sure whether we should expose all this.

After all, a user could create the "schema" manually themselves, using the native APIs. Actually that's what I'd expect to happen in practice anyways, based on my experiences with RDMBS. You always wanted to set some DB-specific magic switches in the DDL, never available through a generic tool such as Hibernate.